### PR TITLE
Cosmos DB Spark: Conversion mismatch configuration

### LINKED ISSUE #20222: Return 'null' when column values cannot be properly transferred using the provided schema. ​
Return 'null' when column values cannot be properly transferred using the provided schema.
The idea is to never lock out the user.

please have a test for the expected behaviour.

### PR DESCRIPTION
**Pull request by ealsur**
When Json documents are read, we do a mapping between the Schema (either provided by the user or inferred) and the Json content.

In the case where the types do not match directly (for example, if the Schema has a LongType and the Json node has a string) we attempt a conversion.

This PR introduces a configuration `spark.cosmos.read.schemaConversionMode` that has 2 modes:

* Relaxed (default): The value for the Row for that attribute will be `null` if the conversion fails
* Strict: Throws exceptions whenever conversion cannot be done

Closes #20222

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/docs/configuration-reference.md [hunk 1]
```diff
@@ -46,6 +46,14 @@ When doing read operations, users can specify a custom schema or allow the conne
 | `spark.cosmos.read.inferSchemaIncludeSystemProperties`     | `false`    | When schema inference is enabled, whether the resulting schema will include all [Cosmos DB system properties](https://docs.microsoft.com/azure/cosmos-db/account-databases-containers-items#properties-of-an-item). |
 | `spark.cosmos.read.inferSchemaIncludeTimestamp`     | `false`    | When schema inference is enabled, whether the resulting schema will include the document Timestamp (`_ts`). Not required if `spark.cosmos.read.inferSchemaIncludeSystemProperties` is enabled, as it will already include all system properties. |
 
+#### Json conversion configuration
+
+When reading json documents, if a document contains an attribute that does not map to the schema type, the user can decide whether to use a `null` value (Relaxed) or an exception (Strict). 
+
+| Config Property Name      | Default | Description |
+| :---        |    :----   |         :--- | 
+| `spark.cosmos.read.jsonSchemaConversionMode`     | `Relaxed`    | The json schema conversion behavior (Relaxed, Strict) |
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/docs/configuration-reference.md:55]: `jsonSchemaConversionMode` why do we need `json` as prefix?

Cosmos DB is a json based. I would drop the prefix unless if there is a specific reason for that?

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala [hunk 2]
```diff
@@ -117,19 +118,54 @@ private object CosmosAccountConfig {
...
+  private val DefaultJsonSchemaConversionMode: JsonSchemaConversionMode = JsonSchemaConversionModes.Relaxed
+
   private val ForceEventualConsistency = CosmosConfigEntry[Boolean](key = "spark.cosmos.read.forceEventualConsistency",
     mandatory = false,
     defaultValue = Some(true),
     parseFromStringFunction = value => value.toBoolean,
     helpMessage = "Makes the client use Eventual consistency for read operations")
 
+  private val JsonSchemaConversion = CosmosConfigEntry[JsonSchemaConversionMode](
+    key = "spark.cosmos.read.jsonSchemaConversionMode",
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:141]: `jsonSchemaConversionMode` why do we need json as prefix?

Cosmos DB is a json based. I would drop the `json` prefix unless if there is a specific reason for that?

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala [hunk 3]
```diff
@@ -117,19 +118,54 @@ private object CosmosAccountConfig {
...
+    parseFromStringFunction = value => validateJsonSchemaConversion(value),
+    helpMessage = "Defines whether to throw on inconsistencies between schema definition and json attribute " +
+      "types (Strict) or to return null values (Relaxed).")
+
+  private def validateJsonSchemaConversion(mode: String): JsonSchemaConversionMode = {
+    Option(mode).fold(DefaultJsonSchemaConversionMode)(p => {
+      val modeName = p.trim
+
+      if (modeName.isEmpty) {
+        DefaultJsonSchemaConversionMode
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:153]: you are already setting the default above `defaultValue = Some(DefaultJsonSchemaConversionMode)`

can we drop the additional logic and rely on the `defaultValue` functionality of the CosmosConfigEntry?

**Review by moderakh — APPROVED** [review]
LGTM. 

a few minor comments.

**Review by FabianMeiswinkel — APPROVED** [review]
Nice! Thanks

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala (hunk 3 above)
Comment (ealsur) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:153]: This handles the fact that the user is specifying the config but its an empty string

Comment (moderakh) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:153]: then we should throw. that's invalid config

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala (hunk 2 above)
Comment (ealsur) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala]: Renamed

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/docs/configuration-reference.md (hunk 1 above)
Comment (ealsur) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/docs/configuration-reference.md]: Renamed

**Pull request merged**